File: src/net/justdave/nwsweatheralertswidget/NWSAlertList.java
package net.justdave.nwsweatheralertswidget;

import java.util.ArrayList;
import java.util.List;

import android.os.Parcel;
import android.os.Parcelable;

public class NWSAlertList extends ArrayList<NWSAlertEntry> implements Parcelable {

    private static final long serialVersionUID = 1L;
    public static final Creator<NWSAlertList> CREATOR = new Creator<NWSAlertList>() {
        @Override
        public NWSAlertList[] newArray(int size) {
            return new NWSAlertList[size];
        }

        @Override
        public NWSAlertList createFromParcel(Parcel source) {
            return new NWSAlertList(source);
        }
    };

    public NWSAlertList() {
        super();
    }

    @SuppressWarnings("unchecked")
    private NWSAlertList(Parcel source) {
        this.clear();
        this.addAll(source.readArrayList(NWSAlertEntry.class.getClassLoader()));
    }

    @Override
    public String toString() {
        String result = "";
        Integer numAlerts = this.size();
        result = result.concat("There are currently ");
        result = result.concat(numAlerts.toString());
        result = result.concat(" active alerts");
        for (NWSAlertEntry entry : this) {
            result = result.concat("\n").concat(entry.toString());
        }
        return result;
    }
    @Override
    public int describeContents() {
        // TODO Auto-generated method stub
        return 0;
    }
    @Override
    public void writeToParcel(Parcel dest, int flags) {
        // TODO Auto-generated method stub
        dest.writeList((List<NWSAlertEntry>) this);
    }

}

File: app/src/main/java/net/justdave/nwsweatheralertswidget/objects/NWSAlertList.kt
package net.justdave.nwsweatheralertswidget.objects

import android.os.Parcel
import android.os.Parcelable
import java.util.*

class NWSAlertList() : ArrayList<NWSAlert>(), Parcelable {

    constructor(parcel: Parcel) : this() {
        arrayListOf<NWSAlert>().apply {
            parcel.readList(this, NWSAlert::class.java.classLoader)
        }
    }

    override fun toString(): String {
        var result = ""
        val numAlerts = size
        result += "There are currently "
        result += numAlerts.toString()
        result = "$result active alerts"
        for (entry in this) {
            result = """
                $result
                $entry
                """.trimIndent()
        }
        return result
    }

    override fun equals(other: Any?): Boolean {
        //formatter:off
        if (this === other) {
            return true
        }
        if (other !is NWSAlertList) {
            return false
        }
        if (size != other.size) {
            return false
        }
        val n = size
        for (i in 0 until n) {
            if (get(i) != other[i]) {
                return false
            }
        }
        //formatter:on
        return true
    }

    @Suppress("EmptyMethod")
    override fun hashCode(): Int {
        return super.hashCode()
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeList(this as List<NWSAlert?>)
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<NWSAlertList> {
        override fun createFromParcel(parcel: Parcel): NWSAlertList {
            return NWSAlertList(parcel)
        }

        override fun newArray(size: Int): Array<NWSAlertList?> {
            return arrayOfNulls(size)
        }
    }

}
